src/Libraries/Microsoft.Extensions.AI/Utilities/AIJsonUtilities.Schema.cs

```diff
@@ -0,0 +1,348 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Concurrent;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Text.Json;
+using System.Text.Json.Nodes;
+using System.Text.Json.Schema;
+using Microsoft.Shared.Diagnostics;
+
+#pragma warning disable S1121 // Assignments should not be made from within sub-expressions
+#pragma warning disable S107 // Methods should not have too many parameters
+#pragma warning disable S1075 // URIs should not be hardcoded
+
+using FunctionParameterKey = (
+    System.Type? Type,
+    string? ParameterName,
+    string? Description,
+    bool HasDefaultValue,
+    object? DefaultValue,
+    bool IncludeSchemaUri,
+    bool DisallowAdditionalProperties,
+    bool IncludeTypeInEnumSchemas);
+
+namespace Microsoft.Extensions.AI;
+
+/// <summary>Provides a collection of utility methods for marshalling JSON data.</summary>
+public static partial class AIJsonUtilities
+{
+    /// <summary>The uri used when populating the $schema keyword in inferred schemas.</summary>
+    private const string SchemaKeywordUri = "https://json-schema.org/draft/2020-12/schema";
+
+    /// <summary>Soft limit for how many items should be stored in the dictionaries in <see cref="_schemaCaches"/>.</summary>
+    private const int CacheSoftLimit = 4096;
+
+    /// <summary>Caches of generated schemas for each <see cref="JsonSerializerOptions"/> that's employed.</summary>
+    private static readonly ConditionalWeakTable<JsonSerializerOptions, ConcurrentDictionary<FunctionParameterKey, JsonElement>> _schemaCaches = new();
+
+    /// <summary>Gets a JSON schema accepting all values.</summary>
+    private static readonly JsonElement _trueJsonSchema = ParseJsonElement("true"u8);
+
+    /// <summary>Gets a JSON schema only accepting null values.</summary>
+    private static readonly JsonElement _nullJsonSchema = ParseJsonElement("""{"type":"null"}"""u8);
+
+    /// <summary>
+    /// Determines a JSON schema for the provided parameter metadata.
+    /// </summary>
+    /// <param name="parameterMetadata">The parameter metadata from which to infer the schema.</param>
+    /// <param name="functionMetadata">The containing function metadata.</param>
+    /// <param name="serializerOptions">The options used to extract the schema from the specified type.</param>
+    /// <param name="inferenceOptions">The options controlling schema inference.</param>
+    /// <returns>A JSON schema document encoded as a <see cref="JsonElement"/>.</returns>
+    public static JsonElement ResolveParameterSchema(
```

Comment:
What's the rationale for ResolveParameterSchema vs ResolveParameterJsonSchema?

Reply:
That's a mistake, I'll see if I can correct it in a follow-up.